### PR TITLE
Allocate linear memory using mmap

### PR DESCRIPTION
**Pull request by loganek**
With this approach we can omit using memset() for the newly allocated memory therefore the physical pages are not being used unless touched by the program.
    
This also simplifies the implementation. There's a room for further refactoring and unifying aot/wasm runtimes, but we can do it as follow-ups to keep the PR at reasonable size.

### CODE REVIEW

**Comment by loganek** [comment]
I need to do a bit more testing before merging the PR, but I think it's in a state where can be reviewed. Any feedback is very welcome.

File: product-mini/platforms/linux-sgx/enclave-sample/Enclave/Enclave.config.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
   <ISVSVN>0</ISVSVN>
   <StackMaxSize>0x100000</StackMaxSize>
   <HeapMaxSize>0x8000000</HeapMaxSize>
-  <ReservedMemMaxSize>0x1000000</ReservedMemMaxSize>
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply product-mini/platforms/linux-sgx/enclave-sample/Enclave/Enclave.config.xml:7]: Normally this change won't break the existing use cases, but it impacts the startup time, I remember we had a test before, the enclave starts slowly when the reserved memory are large.

Maybe you can use `sed` command to automatically replace the string to 0x8000000 when WAMR_BUILD_SPEC_TEST is enabled:
```diff
diff --git a/product-mini/platforms/linux-sgx/CMakeLists.txt b/product-mini/platforms/linux-sgx/CMakeLists.txt
index 0b2b4afe..81d6244b 100644
--- a/product-mini/platforms/linux-sgx/CMakeLists.txt
+++ b/product-mini/platforms/linux-sgx/CMakeLists.txt
@@ -176,3 +176,10 @@ else()
         OUTPUT_VARIABLE cmdOutput
     )
 endif()
+
+if (WAMR_BUILD_SPEC_TEST EQUAL 1)
+    execute_process(
+        COMMAND bash -c "sed -i -E 's/<ReservedMemMaxSize>0x1000000<\\/ReservedMemMaxSize>/<ReservedMemMaxSize>0x8000000<\\/ReservedMemMaxSize>/g' ${CMAKE_CURRENT_SOURCE_DIR}/enclave-sample/Enclave/Enclave.config.xml"
+        OUTPUT_VARIABLE cmdOutput
+    )
+endif()
```
There are already some other samples in the CMakeLists.txt.

File: core/iwasm/common/wasm_memory.c [hunk 2]
```diff
@@ -817,29 +751,64 @@ wasm_enlarge_memory_internal(WASMModuleInstance *module, uint32 inc_page_count)
...
 
+    if (full_size_mmaped) {
 #ifdef BH_PLATFORM_WINDOWS
-    if (!os_mem_commit(memory->memory_data_end,
-                       (uint32)total_size_new - total_size_old,
-                       MMAP_PROT_READ | MMAP_PROT_WRITE)) {
-        ret = false;
-        goto return_func;
-    }
+        if (!os_mem_commit(memory->memory_data_end,
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/common/wasm_memory.c:756]: The `memory->memory_data_end` should be equal to `memory->memory_data + total_size_old`:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L359
`memory->memory_data_end = memory->memory_data + (uint32)memory_data_size;` and `memory_data_size` is `memory_data_size = (uint64)num_bytes_per_page * init_page_count`

But I found there may be an issue in the original wasm_memory.c:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/6daaf6d27a014ecbc4aa67995122f7507a8ea084/core/iwasm/interpreter/wasm_runtime.c#L315-L316
It passes max_memory_data_size to wasm_mmap_linear_memory, which is difference from other two places:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L332-L333
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_runtime.c#L344-L346

File: core/iwasm/common/wasm_memory.c [hunk 3]
```diff
@@ -629,168 +637,94 @@ wasm_runtime_set_mem_bound_check_bytes(WASMMemoryInstance *memory,
...
-            goto return_func;
-        }
-        if (memory_data_old) {
-            bh_memcpy_s(memory_data_new, (uint32)total_size_new,
-                        memory_data_old, total_size_old);
-            wasm_runtime_free(memory_data_old);
-        }
-    }
-
-    memset(memory_data_new + total_size_old, 0,
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/common/wasm_memory.c:718]: For mmap, it is unnecessary to memset it, see:
https://github.com/bytecodealliance/wasm-micro-runtime/commit/84b1a6c10ed5fe395e2f0dce02d99ad3230af024

For wasm_runtime_malloc, had better memset it, but anyway, we are changing to use mmap for all cases.

File: build-scripts/config_common.cmake [hunk 4]
```diff
@@ -248,12 +248,6 @@ if (WAMR_BUILD_SHARED_MEMORY EQUAL 1)
 else ()
   add_definitions (-DWASM_ENABLE_SHARED_MEMORY=0)
 endif ()
-if (WAMR_ENABLE_SHARED_MEMORY_MMAP EQUAL 1)
-  add_definitions (-DWASM_ENABLE_SHARED_MEMORY_MMAP=1)
-  message ("     Shared memory allocated using mmap enabled")
-else ()
-  add_definitions (-DWASM_ENABLE_SHARED_MEMORY_MMAP=0)
```
Comment (wenyongh) [inline build-scripts/config_common.cmake:255]: Can also remove
```C
/* Shared memory */
#ifndef WASM_ENABLE_SHARED_MEMORY
#define WASM_ENABLE_SHARED_MEMORY 0
#endif
```
in core/config.h

File: core/iwasm/aot/aot_runtime.c [hunk 5]
```diff
@@ -519,55 +496,19 @@ memory_instantiate(AOTModuleInstance *module_inst, AOTModuleInstance *parent,
                 module->aux_stack_size);
     LOG_VERBOSE("  heap offset: %u, heap size: %d\n", heap_offset, heap_size);
 
-    memory_data_size = (uint64)num_bytes_per_page * init_page_count;
     max_memory_data_size = (uint64)num_bytes_per_page * max_page_count;
     bh_assert(memory_data_size <= UINT32_MAX);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:500]: memory_data_size isn't set, bh_assert may fail in debug mode

File: core/shared/platform/common/posix/posix_memmap.c [hunk 6]
```diff
@@ -235,6 +240,23 @@ os_munmap(void *addr, size_t size)
     }
 }
 
+#if WASM_HAVE_MREMAP != 0
+void *
+os_mremap(void *old_addr, size_t old_size, size_t new_size)
+{
+    void *ptr = mremap(old_addr, old_size, new_size, MREMAP_MAYMOVE);
+
+    if (ptr == MAP_FAILED) {
```
Comment (wenyongh) [inline core/shared/platform/common/posix/posix_memmap.c:249]: When mremap fails, had better mmap a new memory with new_size, copy old memory data to it and then munmap the old memory.

**Comment by wenyongh** [comment]
@loganek thanks a lot for the refactoring, I tried to review the code and added several comments. Since this is a big change which impacts many platforms (per my understanding, it's a breaking change) and hasn't been tested by developers for some time, could we leave it unmerged until WAMR-1.3.2 is released? I am afraid that it may be a little rushed to merge it because we want to release 1.3.2 soon.

**Comment by loganek** [comment]
> @loganek thanks a lot for the refactoring, I tried to review the code and added several comments. Since this is a big change which impacts many platforms (per my understanding, it's a breaking change) and hasn't been tested by developers for some time, could we leave it unmerged until WAMR-1.3.2 is released? I am afraid that it may be a little rushed to merge it because we want to release 1.3.2 soon.

Absolutely, let's not block 1.3.2 and we can pull it for the next release. And thanks a lot for the review, I'll address your comments in the upcoming days.

File: product-mini/platforms/linux-sgx/enclave-sample/Enclave/Enclave.config.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (loganek) [reply product-mini/platforms/linux-sgx/enclave-sample/Enclave/Enclave.config.xml]: That's a good  idea, thanks!

File: build-scripts/config_common.cmake (hunk 4 above)
Comment (loganek) [reply build-scripts/config_common.cmake:255]: I guess you meant `WASM_ENABLE_SHARED_MEMORY_MMAP` - removed.

File: core/iwasm/aot/aot_runtime.c (hunk 5 above)
Comment (loganek) [reply core/iwasm/aot/aot_runtime.c:500]: right, deleted the assert as we already have the same one when the value is actually set in `wasm_allocate_linear_memory`

File: core/iwasm/common/wasm_memory.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (loganek) [reply core/iwasm/common/wasm_memory.c:756]: > The memory->memory_data_end should be equal to memory->memory_data + total_size_old:

Oh you're right about that.

> But I found there may be an issue in the original wasm_memory.c:

I don't think it's an issue - this code is for shared linear memory only, so mmaping max memory is desired here.

**Comment by loganek** [comment]
@wenyongh I've replied to all your comments, please let me know if you have any question, and feel free to merge the PR once 1.3.2 is relaesed.

File: core/iwasm/common/wasm_memory.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/common/wasm_memory.c:756]: OK, got it, thanks!

File: core/shared/utils/bh_memutils.c [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+void *
+bh_memory_remap_slow(void *old_addr, size_t old_size, size_t new_size)
+{
+    void *new_memory =
+        os_mmap(NULL, new_size, MMAP_PROT_WRITE | MMAP_PROT_READ, 0, -1);
+    if (!new_memory) {
+        return NULL;
+    }
+    bh_memcpy_s(new_memory, new_size, old_addr, old_size);
```
Comment (wenyongh) [inline core/shared/utils/bh_memutils.c:16]: A potential issue is that bh_memcpy_s currently only supports size of uint32, so size may be truncated if it is larger than UINT32_MAX, e.g. in memory64 feature. How about using memcpy and add comment?

**Comment by wenyongh** [comment]
> @wenyongh I've replied to all your comments, please let me know if you have any question, and feel free to merge the PR once 1.3.2 is released.

OK, let's merge it after 1.3.2 and do more tests then.

**Pull request merged**